File: src/test/java/com/fleshgrinder/platform/OsTest.java
package com.fleshgrinder.platform;

import java.io.File;
import java.util.function.Function;
import java.util.function.Supplier;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.EnumSource;
import org.junitpioneer.jupiter.ClearSystemProperty;
import org.junitpioneer.jupiter.SetSystemProperty;

import static com.fleshgrinder.platform.SystemProperties.FS_SEP;
import static com.fleshgrinder.platform.SystemProperties.OS_NAME;
import static com.fleshgrinder.platform.SystemProperties.VM_NAME;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.params.provider.EnumSource.Mode.EXCLUDE;

final class OsTest {
    @ParameterizedTest
    @ClearSystemProperty(key = OS_NAME)
    @ClearSystemProperty(key = FS_SEP)
    @ClearSystemProperty(key = VM_NAME)
    @CsvSource({
        "UNKNOWN, ''",
        "UNKNOWN, '    '",
        "UNKNOWN, '\t\n\t\n'",
        "UNKNOWN, rv32imac",
        "UNKNOWN, rv64imac",
        "UNKNOWN, os4000",
        "AIX, AIX",
        "ANDROID, Android",
        "DARWIN, Darwin",
        "DARWIN, Mac OS X",
        "DARWIN, Mac OS",
        "DRAGONFLYBSD, dragonflybsd",
        "FREEBSD, freebsd",
        "FUCHSIA, fuchsia",
        "HAIKU, haiku",
        "HPUX, hpux",
        "IBMI, ibmi",
        "IBMI, os400",
        "ILLUMOS, illumos",
        "LINUX, linux",
        "NETBSD, netbsd",
        "OPENBSD, openbsd",
        "PLAN9, plan9",
        "QNX, procnto",
        "QNX, qnx",
        "REDOX, redox",
        "SOLARIS, solaris",
        "SOLARIS, sunos",
        "VXWORKS, vxworks",
        "ZOS, zos",

        "AIX, AIX",
        "AIX, garbage aix garbage",
        "ANDROID, ANDROID",
        "ANDROID, garbage android garbage",
        "ANDROID, linux android linux",
        "DARWIN, DARWIN",
        "DARWIN, garbage apple garbage",
        "DARWIN, garbage darwin garbage",
        "DARWIN, garbage mac garbage",
        "DARWIN, garbage mac os garbage",
        "DARWIN, garbage mac os x garbage",
        "DARWIN, garbage mac osx garbage",
        "DARWIN, garbage macos garbage",
        "DARWIN, garbage macos x garbage",
        "DARWIN, garbage macosx garbage",
        "DRAGONFLYBSD, DRAGONFLYBSD",
        "DRAGONFLYBSD, garbage dragonfly garbage",
        "FREEBSD, FREEBSD",
        "FREEBSD, garbage freebsd garbage",
        "FUCHSIA, FUCHSIA",
        "FUCHSIA, garbage fuchsia garbage",
        "HAIKU, garbage haiku garbage",
        "HAIKU, HAIKU",
        "HPUX, garbage hp-ux garbage",
        "HPUX, garbage hpux garbage",
        "HPUX, HPUX",
        "IBMI, garbage ibm i garbage",
        "IBMI, garbage ibmi garbage",
        "IBMI, garbage os400 garbage",
        "IBMI, IBMI",
        "ILLUMOS, garbage illum os garbage",
        "ILLUMOS, garbage illumos garbage",
        "ILLUMOS, ILLUMOS",
        "LINUX, garbage nix garbage",
        "LINUX, garbage nux garbage",
        "LINUX, LINUX",
        "NETBSD, garbage netbsd garbage",
        "NETBSD, NETBSD",
        "OPENBSD, garbage openbsd garbage",
        "OPENBSD, OPENBSD",
        "PLAN9, garbage plan9 garbage",
        "PLAN9, PLAN9",
        "QNX, garbage procnto garbage",
        "QNX, garbage qnx garbage",
        "QNX, QNX",
        "REDOX, garbage redox garbage",
        "REDOX, REDOX",
        "SOLARIS, garbage solaris garbage",
        "SOLARIS, garbage sun os garbage",
        "SOLARIS, garbage sunos garbage",
        "SOLARIS, SOLARIS",
        "VXWORKS, garbage VxWorks garbage",
        "VXWORKS, VXWORKS",
        "WINDOWS, garbage w32 garbage",
        "WINDOWS, garbage w64 garbage",
        "WINDOWS, garbage win32 garbage",
        "WINDOWS, garbage windows garbage",
        "WINDOWS, Windows 10",
        "WINDOWS, Windows 7",
        "WINDOWS, Windows XP",
        "WINDOWS, WINDOWS",
        "ZOS, garbage z/OS garbage",
        "ZOS, garbage zos garbage",
        "ZOS, z/OS",
        "ZOS, ZOS",
    })
    void current(final @NotNull Os expected, final @NotNull String osName) {
        SystemProperties.set(OS_NAME, osName);
        assertEquals(expected, Os.current());
    }

    @SetSystemProperty(key = OS_NAME, value = "linux")
    @SetSystemProperty(key = VM_NAME, value = "Dalvik")
    @ClearSystemProperty(key = FS_SEP)
    @Test void currentAndroid() {
        assertEquals(Os.ANDROID, Os.current());
    }

    @SetSystemProperty(key = FS_SEP, value = "\\")
    @ClearSystemProperty(key = VM_NAME)
    @ClearSystemProperty(key = OS_NAME)
    @Test void currentWindows() {
        assertEquals(Os.WINDOWS, Os.current());
    }

    private static void assertExtension(
        final @NotNull String ext,
        final @NotNull Supplier<String> getExt,
        final @NotNull Function<String, String> withExtString,
        final @NotNull Function<File, File> withExtFile
    ) {
        assertAll(
            () -> assertEquals(ext, getExt.get()),
            () -> assertEquals("some/path" + ext, withExtString.apply("some/path")),
            () -> assertEquals(new File("some/path" + ext), withExtFile.apply(new File("some/path")))
        );
    }

    @Test void executableExtensionWindows() {
        assertExtension(
            ".exe",
            Os.WINDOWS::getExecutableExtension,
            Os.WINDOWS::withExecutableExtension,
            Os.WINDOWS::withExecutableExtension
        );
    }

    @ParameterizedTest
    @EnumSource(value = Os.class, names = {"WINDOWS"}, mode = EXCLUDE)
    void executableExtension(final @NotNull Os os) {
        assertExtension(
            "",
            os::getExecutableExtension,
            os::withExecutableExtension,
            os::withExecutableExtension
        );
    }

    @Test void linkLibraryExtensionWindows() {
        assertExtension(
            ".lib",
            Os.WINDOWS::getLinkLibraryExtension,
            Os.WINDOWS::withLinkLibraryExtension,
            Os.WINDOWS::withLinkLibraryExtension
        );
    }

    @ParameterizedTest
    @EnumSource(value = Os.class, names = {"WINDOWS"}, mode = EXCLUDE)
    void linkLibraryExtension(final @NotNull Os os) {
        assertExtension(
            ".so",
            os::getLinkLibraryExtension,
            os::withLinkLibraryExtension,
            os::withLinkLibraryExtension
        );
    }

    @Test void sharedLibraryExtensionWindows() {
        assertExtension(
            ".dll",
            Os.WINDOWS::getSharedLibraryExtension,
            Os.WINDOWS::withSharedLibraryExtension,
            Os.WINDOWS::withSharedLibraryExtension
        );
    }

    @Test void sharedLibraryExtensionDarwin() {
        assertExtension(
            ".dylib",
            Os.DARWIN::getSharedLibraryExtension,
            Os.DARWIN::withSharedLibraryExtension,
            Os.DARWIN::withSharedLibraryExtension
        );
    }

    @ParameterizedTest
    @EnumSource(value = Os.class, names = {"DARWIN", "WINDOWS"}, mode = EXCLUDE)
    void sharedLibraryExtension(final @NotNull Os os) {
        assertExtension(
            ".so",
            os::getSharedLibraryExtension,
            os::withSharedLibraryExtension,
            os::withSharedLibraryExtension
        );
    }

    @Test void staticLibraryExtensionWindows() {
        assertExtension(
            ".lib",
            Os.WINDOWS::getStaticLibraryExtension,
            Os.WINDOWS::withStaticLibraryExtension,
            Os.WINDOWS::withStaticLibraryExtension
        );
    }

    @ParameterizedTest
    @EnumSource(value = Os.class, names = {"WINDOWS"}, mode = EXCLUDE)
    void staticLibraryExtension(final @NotNull Os os) {
        assertExtension(
            ".a",
            os::getStaticLibraryExtension,
            os::withStaticLibraryExtension,
            os::withStaticLibraryExtension
        );
    }
}

File: src/test/kotlin/com/fleshgrinder/platform/OsTest.kt
package com.fleshgrinder.platform

import com.fleshgrinder.junit.clearSystemProperties
import com.fleshgrinder.junit.enumTestsOf
import com.fleshgrinder.junit.generexWithGarbage
import com.fleshgrinder.junit.testsOf
import com.fleshgrinder.junit.toTests
import com.fleshgrinder.junit.withSystemProperties
import com.fleshgrinder.junit.withSystemProperty
import com.fleshgrinder.platform.Os.*
import java.io.File
import org.intellij.lang.annotations.Language
import org.junit.jupiter.api.Assertions.assertAll
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.parallel.ResourceLock
import org.junit.jupiter.api.parallel.Resources.SYSTEM_PROPERTIES

private class OsTest {
    fun assertAlias(expected: Os, alias: String) =
        clearSystemProperties("file.separator", "java.vm.name") {
            withSystemProperty("os.name", alias) {
                assertAll(
                    { assertEquals(expected, current(), "current") },
                    { if (expected.toString() == alias) assertEquals(expected, fromString(alias), "fromId") else assertThrows<IllegalArgumentException>("fromId") { fromString(alias) } },
                    { assertEquals(expected, parse(alias), "parse") },
                )
            }
        }

    fun aliasTestsOf(expected: Os, @Language("RegExp") p1: String) = generexWithGarbage(p1).toTests { assertAlias(expected, it) }
    fun aliasTestsOf(expected: Os, @Language("RegExp") p1: String, @Language("RegExp") p2: String) = generexWithGarbage(p1, p2).toTests { assertAlias(expected, it) }

    fun assertExtension(ext: String, getExt: () -> String, withExtString: (String) -> String, withExtFile: (File) -> File) =
        assertAll(
            { assertEquals(ext, getExt(), "getExt") },
            { assertEquals("some/path$ext", withExtString("some/path"), "withExtString") },
            { assertEquals(File("some/path$ext"), withExtFile(File("some/path")), "withExtFile") },
        )

    @Test fun `current Os must be known`() {
        assertNotNull(currentOrNull())
    }

    @Test @ResourceLock(SYSTEM_PROPERTIES) fun `current throws if os system property was forcefully removed by the user`() {
        assertThrows<IllegalStateException> { clearSystemProperties("file.separator", "java.vm.name", "os.name") { current() } }
    }

    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun unknown() =
        testsOf("", "os4000", "Roaix", "rv32imac", "rv64imac") {
            clearSystemProperties("file.separator", "java.vm.name") {
                withSystemProperty("os.name", it) {
                    assertAll(
                        { assertThrows<IllegalStateException>("current") { current() } },
                        { assertThrows<IllegalArgumentException>("fromId") { fromString(it) } },
                        { assertThrows<IllegalArgumentException>("parse") { parse(it) } },
                    )
                }
            }
        }

    @Test @ResourceLock(SYSTEM_PROPERTIES) fun `current does not return WINDOWS if file separator is not a backslash`() =
        withSystemProperties("file.separator" to "/", "os.name" to "linux") {
            assertNotEquals(WINDOWS, current())
        }

    @Test @ResourceLock(SYSTEM_PROPERTIES) fun `current returns WINDOWS if file separator is a backslash`() =
        withSystemProperty("file.separator", "\\") {
            assertEquals(WINDOWS, current())
        }

    @Test @ResourceLock(SYSTEM_PROPERTIES) fun `current does not return Android if OS name is not Linux and VM name is not Dalvik`() =
        withSystemProperties("os.name" to "Windows", "java.vm.name" to "HotSpot") {
            assertNotEquals(ANDROID, current())
        }

    @Test @ResourceLock(SYSTEM_PROPERTIES) fun `current return Android if OS name is Linux and VM name is Dalvik`() =
        clearSystemProperties("file.separator") {
            withSystemProperties("java.vm.name" to "Dalvik", "os.name" to "Linux") {
                assertEquals(ANDROID, current())
            }
        }

    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun names() =
        enumTestsOf<Os> {
            val name = it.name
            clearSystemProperties("file.separator", "java.vm.name") {
                withSystemProperty("os.name", name) {
                    assertAll(
                        { assertEquals(it, current(), "current") },
                        { assertThrows<IllegalArgumentException>("fromId") { fromString(name) } },
                        { assertEquals(it, parse(name), "parse") },
                    )
                }
            }
        }

    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun ids() =
        enumTestsOf<Os> {
            val id = it.toString()
            clearSystemProperties("file.separator", "java.vm.name") {
                withSystemProperty("os.name", id) {
                    assertAll(
                        { assertEquals(it, current(), "current") },
                        { assertEquals(it, fromString(id), "fromId") },
                        { assertEquals(it, parse(id), "parse") },
                    )
                }
            }
        }

    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `AIX aliases`() = aliasTestsOf(AIX, "AIX")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `ANDROID aliases`() = aliasTestsOf(ANDROID, "Android")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `DARWIN aliases`() = aliasTestsOf(DARWIN, "(Apple|Darwin|iOS)", "Mac( ?OS( ?X)?)?")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `DRAGONFLYBSD aliases`() = aliasTestsOf(DRAGONFLYBSD, "DragonFly( ?BSD)?")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `FREEBSD aliases`() = aliasTestsOf(FREEBSD, "Free ?BSD")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `FUCHSIA aliases`() = aliasTestsOf(FUCHSIA, "Fuchsia")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `HAIKU aliases`() = aliasTestsOf(HAIKU, "Haiku")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `HPUX aliases`() = aliasTestsOf(HPUX, "HP ?UX")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `IBMI aliases`() = aliasTestsOf(IBMI, "IBM ?i", "OS ?400")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `ILLUMOS aliases`() = aliasTestsOf(ILLUMOS, "Illum( ?OS)?")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `LINUX aliases`() = aliasTestsOf(LINUX, "linux", "u?nix")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `NETBSD aliases`() = aliasTestsOf(NETBSD, "Net ?BSD")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `OPENBSD aliases`() = aliasTestsOf(OPENBSD, "Open ?BSD")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `PLAN9 aliases`() = aliasTestsOf(PLAN9, "Plan ?9")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `QNX aliases`() = aliasTestsOf(QNX, "QNX", "procnto")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `REDOX aliases`() = aliasTestsOf(REDOX, "Redox")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `SOLARIS aliases`() = aliasTestsOf(SOLARIS, "Solaris", "Sun ?OS")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `VXWORKS aliases`() = aliasTestsOf(VXWORKS, "VxWorks")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `WINDOWS aliases`() = aliasTestsOf(WINDOWS, "W(in(dows)?)?(7|8|10|32|64|XP)")
    @TestFactory @ResourceLock(SYSTEM_PROPERTIES) fun `ZOS aliases`() = aliasTestsOf(ZOS, "z/?OS")

    @TestFactory fun exeExt() = enumTestsOf<Os>(WINDOWS) { assertExtension("", it::getExecutableExtension, it::withExecutableExtension, it::withExecutableExtension) }
    @Test fun exeExtWindows() = assertExtension(".exe", WINDOWS::getExecutableExtension, WINDOWS::withExecutableExtension, WINDOWS::withExecutableExtension)

    @TestFactory fun linkLibExt() = enumTestsOf<Os>(WINDOWS) { assertExtension(".so", it::getLinkLibraryExtension, it::withLinkLibraryExtension, it::withLinkLibraryExtension) }
    @Test fun linkLibExtWindows() = assertExtension(".lib", WINDOWS::getLinkLibraryExtension, WINDOWS::withLinkLibraryExtension, WINDOWS::withLinkLibraryExtension)

    @TestFactory fun sharedLibExt() = enumTestsOf<Os>(DARWIN, WINDOWS) { assertExtension(".so", it::getSharedLibraryExtension, it::withSharedLibraryExtension, it::withSharedLibraryExtension) }
    @Test fun sharedLibExtWindows() = assertExtension(".dll", WINDOWS::getSharedLibraryExtension, WINDOWS::withSharedLibraryExtension, WINDOWS::withSharedLibraryExtension)
    @Test fun sharedLibExtDarwin() = assertExtension(".dylib", DARWIN::getSharedLibraryExtension, DARWIN::withSharedLibraryExtension, DARWIN::withSharedLibraryExtension)

    @TestFactory fun staticLibExt() = enumTestsOf<Os>(WINDOWS) { assertExtension(".a", it::getStaticLibraryExtension, it::withStaticLibraryExtension, it::withStaticLibraryExtension) }
    @Test fun staticLibExtWindows() = assertExtension(".lib", WINDOWS::getStaticLibraryExtension, WINDOWS::withStaticLibraryExtension, WINDOWS::withStaticLibraryExtension)
}
